lib/filter_option_parser.rb

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2013-2015 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of version 3 of the GNU General Public License as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+
+# This class takes care of transforming the user-provided filter options into
+# actual Filter objects.
+class FilterOptionParser
+  class <<self
+    def to_filter(command, options, global_options)
+      filter = Filter.from_default_definition(command)
+
+      definitions = skip_files_definitions(options.delete("skip-files"))
+      definitions += exclude_definitions(global_options["exclude"])
+
+      definitions.map! { |definition| definition.gsub("\\@", "@") } # Unescape escaped @s
+      definitions.each do |definition|
+        filter.add_element_filter_from_definition(definition)
+      end
+
+      filter
+    end
+
+    private
+
+    def exclude_definitions(exclude)
+      return [] if !exclude
+
+      filters = exclude.scan(/(@[^,]*)|\"([^,]*?=[^\"]*)\"|([^,]+=[^=]*)$|([^,]+=[^,]*)/).
+        map(&:compact).flat_map do |filter_definition|
+        if filter_definition[0].start_with?("@")
+          filename = File.expand_path(filter_definition[0][1..-1])
+
+          if !File.exists?(filename)
+            raise Machinery::Errors::MachineryError.new(
+              "The filter file '#{filename}' does not exist."
+            )
+          end
+          File.read(filename).lines.map(&:chomp)
+        else
+          filter_definition
+        end
+      end
+
+      filters.reject!(&:empty?) # Ignore empty filters
+      filters
+    end
+
+    def skip_files_definitions(skip_files)
```

Comment:
Assignment Branch Condition size for skip_files_definitions is too high. [15.59/15]